### PR TITLE
Refactor/currency type

### PR DESCRIPTION
**Pull request by sander2**

### CODE REVIEW

File: crates/currency/src/lib.rs [hunk 1]
```diff
@@ -14,23 +15,419 @@ use frame_support::{
...
+    #[pallet::error]
+    pub enum Error<T> {
+        ArithmeticOverflow,
+        ArithmeticUnderflow,
+        TryIntoIntError,
+        InvalidCurrency,
+    }
+
+    #[pallet::pallet]
+    pub struct Pallet<T>(_);
```
Comment (gregdhill) [inline crates/currency/src/lib.rs:88]: Perhaps I'm missing something, but why does this need to be a pallet if there are no storage items?

File: crates/currency/src/lib.rs [hunk 2]
```diff
@@ -14,23 +15,419 @@ use frame_support::{
...
+        pub const fn new(amount: BalanceOf<T>, currency_id: CurrencyId<T>) -> Self {
+            Self { amount, currency_id }
+        }
+
+        pub fn zero(currency_id: CurrencyId<T>) -> Self {
+            Self::new(0u32.into(), currency_id)
+        }
+
+        pub fn from_fixed_point(
+            amount: SignedFixedPoint<T>,
```
Comment (gregdhill) [inline crates/currency/src/lib.rs:118]: Why signed and not unsigned? Perhaps we should change the name of the function?

File: crates/currency/src/lib.rs [hunk 3]
```diff
@@ -14,23 +15,419 @@ use frame_support::{
...
+        }
+
+        pub fn checked_div(&self, scalar: &UnsignedFixedPoint<T>) -> Result<Self, DispatchError> {
+            let amount = UnsignedFixedPoint::<T>::checked_from_integer(self.amount)
+                .ok_or(Error::<T>::TryIntoIntError)?
+                .checked_div(&scalar)
+                .ok_or(Error::<T>::ArithmeticOverflow)?
+                .into_inner()
+                .checked_div(&UnsignedFixedPoint::<T>::accuracy())
+                .ok_or(Error::<T>::ArithmeticUnderflow)?;
```
Comment (gregdhill) [inline crates/currency/src/lib.rs:225]: Should we extract this logic somewhere? There are a few places we want to divide the inner by the accuracy.

File: crates/currency/src/lib.rs [hunk 4]
```diff
@@ -14,23 +15,419 @@ use frame_support::{
...
+        }
+
+        pub fn ratio(&self, other: &Self) -> Result<UnsignedFixedPoint<T>, DispatchError> {
+            ensure!(self.currency_id == other.currency_id, Error::<T>::InvalidCurrency);
+            let ratio = UnsignedFixedPoint::<T>::checked_from_rational(self.amount, other.amount)
+                .ok_or(Error::<T>::TryIntoIntError)?;
+            Ok(ratio)
+        }
+
+        pub fn to_fixed(&self) -> Result<SignedFixedPoint<T>, DispatchError> {
```
Comment (gregdhill) [inline crates/currency/src/lib.rs:239]: As above, maybe make the function name more explicit?

File: crates/currency/src/lib.rs [hunk 5]
```diff
@@ -14,23 +15,419 @@ use frame_support::{
...
+
+        pub fn ge(&self, other: &Self) -> Result<bool, DispatchError> {
+            ensure!(self.currency_id == other.currency_id, Error::<T>::InvalidCurrency);
+            Ok(self.amount >= other.amount)
+        }
+
+        pub fn gt(&self, other: &Self) -> Result<bool, DispatchError> {
+            ensure!(self.currency_id == other.currency_id, Error::<T>::InvalidCurrency);
+            Ok(self.amount > other.amount)
+        }
```
Comment (gregdhill) [inline crates/currency/src/lib.rs:275]: Maybe we should add a `CheckedPartialOrd` trait for these?

File: crates/currency/src/lib.rs [hunk 6]
```diff
@@ -14,23 +15,419 @@ use frame_support::{
...
+            with_currency_id::mint::<T>(self.currency_id, account_id, self.amount)
+        }
+
+        // lock, unlock, etc
+
+        pub fn convert_to(&self, currency_id: CurrencyId<T>) -> Result<Self, DispatchError> {
+            T::CurrencyConversion::convert(self, currency_id)
+        }
+
+        pub fn is_zero(&self) -> bool {
```
Comment (gregdhill) [inline crates/currency/src/lib.rs:305]: We should implement `Zero` in `num-traits`.

File: crates/exchange-rate-oracle/src/mock.rs [hunk 7]
```diff
@@ -19,13 +20,16 @@ frame_support::construct_runtime!(
         NodeBlock = Block,
         UncheckedExtrinsic = UncheckedExtrinsic,
     {
+        Tokens: orml_tokens::{Pallet, Storage, Config<T>, Event<T>},
```
Comment (gregdhill) [inline crates/exchange-rate-oracle/src/mock.rs:23]: Minor nit, put this under the core Substrate pallets.

File: crates/currency/src/tests.rs [hunk 8]
```diff
@@ -0,0 +1,39 @@
...
+            ),
+            (
+                Amount::new(1 * 10u128.pow(10), currency),// 1 DOT
+                FixedU128::checked_from_rational(1, 10).unwrap(), // 10%
+                Amount::new(1000000000, currency),
+            ),
+        ];
+
+        for (amount, percent, expected) in tests {
+            let actual = Fee::calculate_for(amount, percent).unwrap();
```
Comment (gregdhill) [inline crates/currency/src/tests.rs:35]: Looks like you removed `calculate_for` in the fee pallet? Not sure why this compiles...

Comment (gregdhill) [reply crates/currency/src/tests.rs:35]: In any case I think we should take this opportunity to rename `calculate_for` to something more explicit.

File: crates/fee/src/lib.rs [hunk 9]
```diff
@@ -444,24 +418,9 @@ impl<T: Config> Pallet<T> {
...
-            SignedFixedPoint::checked_from_integer(reward_as_inner).ok_or(Error::<T>::TryIntoIntError)?;
-        Ok(Rewards::distribute_reward(reward_as_fixed)?
-            .into_inner()
-            .checked_div(&SignedFixedPoint::accuracy())
-            .ok_or(Error::<T>::ArithmeticUnderflow)?
-            .try_into()
-            .ok()
-            .ok_or(Error::<T>::TryIntoIntError)?)
+    fn distribute(reward: &Amount<T>) -> Result<Amount<T>, DispatchError> {
+        let leftover = T::VaultRewards::distribute_reward(reward.to_fixed()?)?;
```
Comment (gregdhill) [inline crates/fee/src/lib.rs:422]: Should we verify here that the `CurrencyId` of the `Amount` is the rewards currency?

File: crates/replace/src/lib.rs [hunk 10]
```diff
@@ -295,72 +296,69 @@ impl<T: Config> Pallet<T> {
...
+                &replace_collateral_increase,
             )?;
 
         // check that total-to-be-replaced is above the minimum. NOTE: this means that even
         // a request with amount=0 is valid, as long the _total_ is above DUST. This might
         // be the case if the vault just wants to increase the griefing collateral, for example.
-        let dust_value = <ReplaceBtcDustValue<T>>::get();
-        ensure!(total_to_be_replaced >= dust_value, Error::<T>::AmountBelowDustAmount);
+        ensure!(
+            total_to_be_replaced.ge(&Self::dust_value())?,
```
Comment (gregdhill) [inline crates/replace/src/lib.rs:327]: Feels weird to me that this operation is fallible, i.e. the dust value should always have the same `CurrencyId` but we now have a runtime error if for some reason these are out-of-sync.

**Review by gregdhill — DISMISSED** [review]
First pass, still need to review some of the files with bigger diffs. Looks good overall!

File: crates/currency/src/lib.rs (hunk 1 above)
Comment (sander2) [reply crates/currency/src/lib.rs:88]: For the error type

File: crates/currency/src/lib.rs (hunk 6 above)
Comment (sander2) [reply crates/currency/src/lib.rs:305]: `Zero` requires a `fn zero() -> Self` implementation, which due to the `currency_id` does not really make sense

File: crates/currency/src/lib.rs (hunk 5 above)
Comment (sander2) [reply crates/currency/src/lib.rs:275]: Why? Imo that only makes sense if we implement this trait also for other types

File: crates/currency/src/lib.rs (hunk 3 above)
Comment (sander2) [reply crates/currency/src/lib.rs:225]: Extracted into a `truncate_to_inner` trait implementation

File: crates/currency/src/tests.rs (hunk 8 above)
Comment (sander2) [reply crates/currency/src/tests.rs:35]: Woops, the file was not being included. `calculate_for` was replace by the `checked_fixed_point_mul` method on Amount.

File: crates/currency/src/lib.rs (hunk 1 above)
Comment (gregdhill) [reply crates/currency/src/lib.rs:88]: We don't need a pallet for the `Error` type, look at the `bitcoin` crate for example

File: crates/replace/src/lib.rs (hunk 10 above)
Comment (sander2) [reply crates/replace/src/lib.rs:327]: Well, we also use checked math even when we are sure that the math can not fail, this is not much different. We could compare the `.amount()` values directly, but I think that we should probably keep amount comparisons consistent across the code. Maybe in the future we can extend the type to be able to statically distinguish `Wrapped` and `Collateral` amounts, but I don't think it's very urgent.

File: crates/currency/src/lib.rs (hunk 1 above)
Comment (sander2) [reply crates/currency/src/lib.rs:97]: But then you need error conversions everywhere, plus you lose information that the error occurred in the `currency` crate

File: crates/currency/src/lib.rs [hunk 11]
```diff
@@ -13,24 +20,422 @@ use frame_support::{
...
+#[cfg_attr(feature = "testing-utils", mocktopus::macros::mockable)]
+mod monetary {
+    use super::*;
+
+    #[cfg_attr(feature = "testing-utils", derive(Copy))]
+    #[derive(Clone, PartialEq, Eq, Debug)]
+    pub struct Amount<T: Config> {
+        amount: BalanceOf<T>,
+        currency_id: CurrencyId<T>,
+    }
```
Comment (gregdhill) [inline crates/currency/src/lib.rs:115]:
```suggestion
    pub struct Amount<Balance, CurrencyId> {
        amount: Balance,
        currency_id: CurrencyId,
    }
```

Comment (gregdhill) [reply crates/currency/src/lib.rs:115]: No need to bind the pallet `Config` to the generic type here.

File: crates/currency/src/lib.rs [hunk 12]
```diff
@@ -13,24 +20,422 @@ use frame_support::{
...
+            ensure!(self.currency_id == other.currency_id, Error::<T>::InvalidCurrency);
+            Ok(self.amount >= other.amount)
+        }
+
+        pub fn gt(&self, other: &Self) -> Result<bool, DispatchError> {
+            ensure!(self.currency_id == other.currency_id, Error::<T>::InvalidCurrency);
+            Ok(self.amount > other.amount)
+        }
+
+        // todo: all other checked_ math operations
```
Comment (gregdhill) [inline crates/currency/src/lib.rs:286]: Are they needed or not?

File: crates/currency/src/lib.rs [hunk 13]
```diff
@@ -13,24 +20,422 @@ use frame_support::{
...
+            // note that unwrapping is safe because we use a constant
+            let rounding_addition = UnsignedFixedPoint::<T>::checked_from_rational(1, 2).unwrap();
+
+            let amount = UnsignedFixedPoint::<T>::checked_from_integer(self.amount)
+                .ok_or(Error::<T>::ArithmeticOverflow)?
+                .checked_mul(&fraction)
+                .ok_or(Error::<T>::ArithmeticOverflow)?
+                .checked_add(&rounding_addition)
+                .ok_or(Error::<T>::ArithmeticOverflow)?
+                .truncate_to_inner()
```
Comment (gregdhill) [inline crates/currency/src/lib.rs:328]: I much prefer this new trait :)

**Review by nud3l — APPROVED** [review]

**Pull request merged**